test/common/json/json_sanitizer_test_util.cc

```diff
@@ -0,0 +1,146 @@
+  // exclusion set or the 4-byte utf-8 exclusion-set.
+  bool isInvalid3Byte(uint32_t unicode) const { return invalid_3byte_intervals_.test(unicode); }
+  bool isInvalid4Byte(uint32_t unicode) const { return invalid_4byte_intervals_.test(unicode); }
+
+private:
+  IntervalSetImpl<uint32_t> invalid_3byte_intervals_;
+  IntervalSetImpl<uint32_t> invalid_4byte_intervals_;
+};
+
+const InvalidUnicodeSet& invalidUnicodeSet() { CONSTRUCT_ON_FIRST_USE(InvalidUnicodeSet); }
+
+} // namespace
+
+bool isProtoSerializableUtf8(absl::string_view in) {
+  const uint8_t* data = reinterpret_cast<const uint8_t*>(in.data());
+  uint32_t size = in.size();
+  while (size != 0) {
+    if ((*data & 0x80) == 0) {
+      ++data;
+      --size;
+    } else {
+      auto [unicode, consumed] = Utf8::decode(data, size);
+      data += consumed;
+      size -= consumed;
+
+      switch (consumed) {
+      case 2:
+        break;
+      case 3:
+        if (invalidUnicodeSet().isInvalid3Byte(unicode)) {
+          return false;
+        }
+        break;
+      case 4:
+        if (invalidUnicodeSet().isInvalid4Byte(unicode)) {
+          return false;
+        }
+        break;
+      default:
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Decodes unicode hex escape \u1234 into 0x1234, returning success.
+bool parseUnicode(absl::string_view str, uint32_t& hex_value) {
+  if (absl::StartsWith(str, "\\u") && str.size() >= 6) {
+    // TODO(jmarantz): Github master,
+    // https://github.com/abseil/abseil-cpp/blob/master/absl/strings/numbers.h
+    // has absl::SimpleHexAtoi, enabling this impl to be
+    //   return absl::SimpleHexAtoi(str.substr(2, 4), &hex_value);
+    // In the meantime we must nul-terminate.
+    std::string hex_str(str.data() + 2, 4);
+    char* str_end;
+    hex_value = strtoul(hex_str.c_str(), &str_end, 16);
+    return str_end != nullptr && *str_end == '\0';
+  }
+  return false;
+}
+
+// Compares a string that's possibly an escaped unicode, e.g. \u1234, to
+// one that is utf8-encoded.
+bool compareUnicodeEscapeAgainstUtf8(absl::string_view& escaped, absl::string_view& utf8) {
+  uint32_t escaped_unicode;
+  if (parseUnicode(escaped, escaped_unicode)) {
+    // If one side of the comparison is a unicode escape,
+    auto [unicode, consumed] = Utf8::decode(utf8);
+    if (consumed != 0 && unicode == escaped_unicode) {
+      utf8 = utf8.substr(consumed, utf8.size() - consumed);
+      escaped = escaped.substr(6, escaped.size() - 6);
+      return true;
+    }
+  }
+  return false;
+}
+
+// Determines whether two strings differ only in whether they have
+// literal utf-8 or escaped 3-byte unicode. We do this equivalence
+// comparison to enable differential fuzzing between JsonSanitizer and
+// protobuf json serialization. The protobuf implementation has made
+// some hard-to-understand decisions about what to encode via unicode
+// escapes versus what to pass through as utf-8.
```

Comment:
Is nlohmann always stricter? If so, then the validation logic should be asymmetric to reflect.